libs/pubsub/pubsub.go

```diff
@@ -195,333 +181,207 @@ func (s *Server) SubscribeUnbuffered(ctx context.Context, clientID string, query
 }
 
 func (s *Server) subscribe(ctx context.Context, clientID string, query Query, outCapacity int) (*Subscription, error) {
-	s.mtx.RLock()
-	clientSubscriptions, ok := s.subscriptions[clientID]
-	if ok {
-		_, ok = clientSubscriptions[query.String()]
-	}
-	s.mtx.RUnlock()
-	if ok {
+	s.subs.Lock()
+	defer s.subs.Unlock()
+
+	if s.subs.index == nil {
+		return nil, ErrServerStopped
+	} else if s.subs.index.contains(clientID, query.String()) {
 		return nil, ErrAlreadySubscribed
 	}
 
-	subscription := NewSubscription(outCapacity)
-	select {
-	case s.cmds <- cmd{op: sub, clientID: clientID, query: query, subscription: subscription}:
-		s.mtx.Lock()
-		if _, ok = s.subscriptions[clientID]; !ok {
-			s.subscriptions[clientID] = make(map[string]string)
-		}
-		s.subscriptions[clientID][query.String()] = subscription.id
-		s.subscriptions[clientID][subscription.id] = query.String()
-		s.mtx.Unlock()
-		return subscription, nil
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	case <-s.Quit():
-		return nil, nil
-	}
+	sub := NewSubscription(outCapacity)
```

Comment:
should we, as part of this forbid unbuffered capacities? 

My current belief is that the legacy implementation's most impacting flaw is allowing unbuffered subscriptions (at all).